packages/docs-infra/src/CodeHighlighter/CodeHighlighter.tsx

```diff
@@ -0,0 +1,520 @@
+  const errors: Error[] = [];
+  for (const item of variantCodes) {
+    if ('error' in item) {
+      errors.push(item.error);
+    } else {
+      processedCode[item.name] = item.variant.code;
+    }
+  }
+
+  if (errors.length > 0) {
+    return <ErrorHandler errors={errors} />;
+  }
+
+  const clientProps = createClientProps({
+    ...props,
+    code: processedCode,
+    processedGlobalsCode,
+  });
+
+  return <CodeHighlighterClient {...clientProps} />;
+
+  // TODO: we might not need the client if hydrateAt is 'init' or 'stream' and there is no setCode() or setSelection()
+}
+
+function renderCodeHighlighter<T extends {}>(
+  props: BaseHelperProps<T> & {
+    fallback?: React.ReactNode;
+    skipFallback?: boolean;
+    processedGlobalsCode?: Array<Code>;
+  },
+) {
+  const ErrorHandler = props.ErrorHandler || CodeErrorHandler;
+
+  const code = props.code || props.precompute;
+  const variants = props.variants || Object.keys(props.components || code || {});
+  const allCodeVariantsLoaded = code && hasAllVariants(variants, code, true);
+
+  if (!allCodeVariantsLoaded) {
+    if (props.forceClient) {
+      return (
+        <ErrorHandler errors={[new Error('Client only mode requires precomputed source code')]} />
+      );
+    }
+
+    return <CodeSourceLoader {...props} />;
+  }
+
+  const clientProps = createClientProps({
+    ...props,
+  });
+
+  return <CodeHighlighterClient {...clientProps} />;
+
+  // TODO: we might not need the client if hydrateAt is 'init' or 'stream' and there is no props.controlled
+}
+
+/**
+ * Ensures that the suspense boundary is always rendered, even if none of the children have async operations.
+ */
+async function CodeHighlighterSuspense(props: { children: React.ReactNode }) {
+  return props.children;
+}
+
+function renderWithInitialSource<T extends {}>(
+  props: BaseHelperProps<T> & {
+    code: Code;
+    initialVariant: string;
+    initialFilename: string | undefined;
+    initialSource: VariantSource;
+    initialExtraFiles?: VariantExtraFiles;
+    ContentLoading: React.ComponentType<ContentLoadingProps<T>>;
+    processedGlobalsCode?: Array<Code>;
+  },
+) {
+  const fileNames = [
+    ...(props.initialFilename ? [props.initialFilename] : []),
+    ...Object.keys(props.initialExtraFiles || {}),
+  ];
+  const source = stringOrHastToJsx(props.initialSource, props.highlightAt === 'init');
+
+  const contentProps = {
+    name: props.name,
+    slug: props.slug,
+    url: props.url,
+    fileNames,
+    source,
+    ...(props.contentProps || {}),
+  } as ContentLoadingProps<T>;
+
+  const ContentLoading = props.ContentLoading;
+  const fallback = <ContentLoading {...contentProps} />;
+
+  if (props.forceClient) {
+    return renderCodeHighlighter({
+      ...props,
+      fallback,
+      processedGlobalsCode: props.processedGlobalsCode,
+    });
+  }
+
+  return (
+    <React.Suspense fallback={fallback}>
+      <CodeHighlighterSuspense>
+        {renderCodeHighlighter({
+          ...props,
```

Comment:
Why not having the spread be last?

Reply:
So that we override any props the parent might have set with the fallback component we generated